situp-plugins/apmtracesource/src/main/java/com/amazon/situp/plugins/processor/oteltrace/RawTrace.java

```diff
@@ -0,0 +1,156 @@
+        private String serviceName;
+        private Map<String, String> resourceAttributes;
+
+        public RawTraceBuilder() {
+        }
+
+        public RawTraceBuilder setTraceId(final String traceId) {
+            checkNotNull(traceId, "traceId cannot be null");
+            this.traceId = traceId;
+            return this;
+        }
+
+        public RawTraceBuilder setSpanId(final String spanId) {
+            checkNotNull(spanId, "spanId cannot be null");
+            this.spanId = spanId;
+            return this;
+        }
+
+        public RawTraceBuilder setParentSpanId(final String parentSpanId) {
+            this.parentSpanId = parentSpanId;
+            return this;
+        }
+
+        public RawTraceBuilder setName(final String name) {
+            this.name = name;
+            return this;
+        }
+
+        public RawTraceBuilder setStartTime(final String startTime) {
+            this.startTime = startTime;
+            return this;
+        }
+
+        public RawTraceBuilder setEndTime(final String endTime) {
+            this.endTime = endTime;
+            return this;
+        }
+
+        public RawTraceBuilder setInstrumentationName(final String instrumentationName) {
+            this.instrumentationName = instrumentationName;
+            return this;
+        }
+
+        public RawTraceBuilder setInstrumentationVersion(final String instrumentationVersion) {
+            this.instrumentationVersion = instrumentationVersion;
+            return this;
+        }
+
+        public RawTraceBuilder setServiceName(final String serviceName) {
+            this.serviceName = serviceName;
+            return this;
+        }
+
+        public RawTraceBuilder setResourceAttributes(final Map<String, String> resourceAttributes) {
+            this.resourceAttributes = resourceAttributes;
+            return this;
+        }
+
+        public RawTrace build() {
+            return new RawTrace(this);
+        }
+
+    }
+
+    private static String convertStringNanosToISO8601(final String stringNanos) {
+        final BigDecimal nanos = new BigDecimal(stringNanos);
+        final long epochSeconds = nanos.divide(MILLIS_TO_NANOS.multiply(SEC_TO_MILLIS), RoundingMode.DOWN).longValue();
+        final int nanoAdj = nanos.remainder(MILLIS_TO_NANOS.multiply(SEC_TO_MILLIS)).intValue();
+        return Instant.ofEpochSecond(epochSeconds, nanoAdj).toString();
+    }
+
+
+    public static RawTrace buildFromProto(Resource resource, Span spans, InstrumentationLibrary instrumentationLibrary) {
+        final String SERVICE_NAME = "service.name";
+
+        Map<Object, Object> map = resource.getAttributesList().stream().collect(Collectors.toMap(keyVal -> keyVal.getKey(), keyVal -> keyVal.getValue()));
+        final Map<String, String> resourceAttributesMap = new HashMap<>();
+        for (Map.Entry<Object, Object> entry : map.entrySet()) {
+            if (entry.getValue() instanceof String) {
```

Comment:
do you expect all of the values to be instances of String here ?

what happens when there is a value like `new Boolean(true)` or `new Integer(23)` - we may end up ignoring those - are we good with it ?

Reply:
Makes sense, let me modify this to convert those objects to String.

Reply:
I think this is redundant now, probably think of doing it like below 

```
//if keyValue has null key, it will be mapped as null -> <whatever is the value>
        Map<String, String> resourceMap = resource.getAttributesList().stream()
                .collect(Collectors.toMap(
                        keyValue -> (String) keyValue.getKey(), 
                        keyValue-> String.valueOf(keyValue.getValue())));
```

or

```
       resource.getAttributesList().forEach( keyValue -> resourceMap.put((String) keyValue.getKey(), String.valueOf(keyValue.getValue())));
          
        });
```

or


```
        resource.getAttributesList().forEach( keyValue -> {
            if(keyValue.getKey() != null) {
                resourceMap.put((String) keyValue.getKey(), String.valueOf(keyValue.getValue()));
            }
        });
```
from the documentation of attributes, it does not look like there can be null keys - so u can ignore the null check of the key and just make the conversion. Also from the documentation it looks like the values are as below so we can safely assume all of the value objects have `toString()` implemented

```
// attributes is a collection of key/value pairs. The value can be a string,
  // an integer, a double or the Boolean values `true` or `false`. Note, global attributes
  // like server name can be set using the resource API. Examples of attributes:
  //
  //     "/http/user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
  //     "/http/server_latency": 300
  //     "abc.com/myattribute": true
  //     "abc.com/score": 10.239
  repeated opentelemetry.proto.common.v1.KeyValue attributes = 9;
```

Reply:
Seems like it can be `AnyValue` https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/common/v1/common.proto#L30-L36 - u can ignore the `String.valueOf()` completely because it can also be Array and KeyValueList